zkvm/docs/zkvm-spec.md

```diff
@@ -147,44 +147,57 @@ transaction’s applicability to the blockchain.
 The items on the ZkVM stack are typed. The available types fall into two 
 categories: [copyable types](#copyable-types) and [linear types](#linear-types).
 
+
 ### Copyable types
 
-Copyable types can be freely created, copied ([`dup`](#dup)), and destroyed ([`drop`](#drop)).
+Copyable types can be freely created, copied ([`dup`](#dup)), and destroyed ([`drop`](#drop)):
 
 * [Data](#data-type)
 * [Variable](#variable-type)
-* [Expression](#expression-type)
-* [Constraint](#constraint-type)
+
+Note: the [program type](#program-type) is not copyable to avoid denial-of-service attacks
+via repeated execution of the same program that can be scaled exponentially while
+growing the transaction size only linearly.
 
 
 ### Linear types
 
 Linear types are subject to special rules as to when and how they may be created
 and destroyed, and may never be copied.
 
+* [Program](#program-type)
+* [Expression](#expression-type)
+* [Constraint](#constraint-type)
 * [Contract](#contract-type)
 * [Wide value](#wide-value-type)
 * [Value](#value-type)
 
 
 ### Portable types
 
-Only the [data](#data-type) and [value](#value-type) types can be _ported_ across transactions via [outputs](#output-structure).
+Only the [data](#data-type), [program](#program-type) and [value](#value-type) types can be _ported_ across transactions via [outputs](#output-structure).
 
 Notes:
 
 * [Wide values](#wide-value-type) are not portable because they are not proven to be non-negative.
 * [Contracts](#contract-type) are not portable because they must be satisfied within the current transaction
 or [output](#output-structure) their contents themselves.
 * [Variables](#variable-type), [expressions](#expression-type) and [constraints](#constraint-type) have no meaning outside the VM state
-and its constraint system and therefore cannot be meaningfully ported between transactions.
+and its constraint system, and therefore cannot be meaningfully ported between transactions.
 
 
 ### Data type
 
-A _data type_ is a variable-length byte array used to represent signatures, proofs and programs.
+A _data type_ is a variable-length byte array used to represent [commitments](#pedersen-commitment), [scalars](#scalar), signatures and proofs.
```

Comment:
Repeating some feedback from an earlier round of review on this spec: "type" is being used in a confusing way here. You can talk about "the data type" or "an instance of type data" but not "a data type." Consider:

> An object of the _data_ type is a variable-length byte array...

(Same thing with "program type" below.)

Reply:
How about renaming it into `String` in some other PR?

Reply:
Sounds great. :+1: